src/core/cfg.cpp

```diff
@@ -2274,105 +2222,197 @@ class CFGVisitor : public ASTVisitor {
         pushAssign(exitname, exit);
-        enter = remapExpr(makeCall(enter));
 
-        if (node->optional_vars) {
+        AST_expr* enter = makeLoadAttribute(makeLoad(ctxmgrname, node), internString("__enter__"), true);
+        enter = remapExpr(makeCall(enter));
+        if (node->optional_vars)
             pushAssign(node->optional_vars, enter);
-        } else {
+        else
             push_back(makeExpr(enter));
-        }
-
-        CFGBlock* continue_dest = NULL, * break_dest = NULL;
-        if (regions.size()) {
-            continue_dest = cfg->addDeferredBlock();
-            continue_dest->info = "with_continue";
-            break_dest = cfg->addDeferredBlock();
-            break_dest->info = "with_break";
 
-            pushLoopRegion(continue_dest, break_dest);
-        }
+        // push continuations
+        CFGBlock* finally_block = cfg->addDeferredBlock();
+        finally_block->info = "with_finally";
+        pushFinallyContinuation(finally_block, whyname);
 
-        CFGBlock* return_dest = cfg->addDeferredBlock();
-        return_dest->info = "with_return";
-        pushReturnRegion(return_dest);
+        CFGBlock* exc_block = cfg->addDeferredBlock();
+        exc_block->info = "with_exc";
+        exc_handlers.push_back({ exc_block, exc_type_name, exc_value_name, exc_traceback_name });
 
         for (int i = 0; i < node->body.size(); i++) {
             node->body[i]->accept(this);
         }
 
-        popRegion(); // for the retrun
-
-        AST_Call* exit_call = makeCall(makeName(exitname, AST_TYPE::Load, node->lineno));
-        exit_call->args.push_back(makeName(nonename, AST_TYPE::Load, node->lineno));
-        exit_call->args.push_back(makeName(nonename, AST_TYPE::Load, node->lineno));
-        exit_call->args.push_back(makeName(nonename, AST_TYPE::Load, node->lineno));
-        push_back(makeExpr(exit_call));
+        exc_handlers.pop_back();
+        int finally_did_why = continuations.back().did_why;
+        popContinuation();
 
-        CFGBlock* orig_ending_block = curblock;
+        if (curblock) {
+            // The try-suite finished as normal; jump to the finally block.
+            pushAssign(whyname, makeNum(Why::FALLTHROUGH));
+            pushJump(finally_block);
+        }
+
+        // This tells us whether our exit_block will have multiple incoming edges (one from exc_block, one from
+        // finally_block) or not (if exc_block and/or finally_block is unneeded, or if finally_block doesn't jump to
+        // exit_block because it's never entered except via continue, break, and/or return). If exit_block has multiple
+        // incoming edges, then anybody *branching* into it (rather than jumping directly) needs to do so through a
+        // "trampoline" block to avoid critical edges.
+        // bool exit_multi_incoming = exc_block->predecessors.size() && (finally_did_why & (1 << Why::FALLTHROUGH));
+        bool exit_multi_incoming = true; // FIXME FIXME FIXME(rntz)
```

Comment:
We also have a unneccessary-block-cleanup phase at the end, so while it's nice if we can avoid generating them in the first place, I think it might be ok to just always generate the trampolines?